packages/optimizer/lib/transformers/OptimizeAssets.js

```diff
@@ -151,28 +152,84 @@ class OptimizeImages {
     this.maxSrcsetValues = MAX_SRCSET_VALUE_COUNT;
   }
 
-  async transform(root) {
-    if (!this.imageOptimizer) {
-      return;
-    }
+  async transform(root, runtimeParams) {
     const html = firstChildByTag(root, 'html');
     const body = firstChildByTag(html, 'body');
+    const {baseDomain, rewriteUrlsToAmpCache} = runtimeParams;
 
     let node = body;
     const imageOptimizationPromises = [];
-    while (node !== null) {
+    while (node) {
       if (node.tagName === 'template') {
         node = skipNodeAndChildren(node);
       } else {
-        if (node.tagName === 'amp-img') {
+        if (node.tagName === 'amp-img' && this.imageOptimizer) {
           imageOptimizationPromises.push(this.optimizeImage(node));
         }
+        if (rewriteUrlsToAmpCache) {
+          imageOptimizationPromises.push(
+            this.rewriteHosts(node, rewriteUrlsToAmpCache, baseDomain)
+          );
+        }
+
         node = nextNode(node);
       }
     }
     return Promise.all(imageOptimizationPromises);
   }
 
+  async _replaceUrl(url, host) {
+    return createCacheUrl(host, url);
+  }
+
+  async rewriteHost(src, host, baseDomain) {
+    if (!src) {
+      return;
+    }
+    if (src.startsWith('/')) {
+      src = baseDomain + src;
+    }
+    return await this._replaceUrl(src, host);
+  }
+
+  async rewriteHosts(node, host, baseDomain) {
+    const tag = node.tagName;
+
+    if (['amp-img', 'img', 'amp-anim'].includes(tag)) {
+      if (hasAttribute(node, 'src')) {
+        node.attribs.src = await this.rewriteHost(node.attribs.src, host, baseDomain);
+      }
+
+      if (hasAttribute(node, 'srcset')) {
+        const srcset = parseSrcSet(node.attribs.srcset);
+        for (const source of srcset.sources_) {
+          source.url = await this.rewriteHost(source.url, host, baseDomain);
+        }
+        // Drop ending comma
+        node.attribs.srcset = srcset.stringify();
+      }
+    } else if (['video', 'amp-video'].includes(tag)) {
+      if (hasAttribute(node, 'poster')) {
+        node.attribs.poster = await this.rewriteHost(node.attribs.poster, host, baseDomain);
+      }
+    } else if (['link'].includes(tag) && node.attribs.rel === 'icon' && node.attribs.href) {
+      node.attribs.href = await this.rewriteHost(node.attribs.href, host, baseDomain);
+    } else if (['image'].includes(tag)) {
+      node.attribs.href = await this.rewriteHost(node.attribs.href, host, baseDomain);
+      node.attribs['xlink:href'] = await this.rewriteHost(
+        node.attribs['xlink:href'],
+        host,
+        baseDomain
+      );
+    } else if (['use'].includes(tag)) {
+      node.attribs['xlink:href'] = await this.rewriteHost(
+        node.attribs['xlink:href'],
+        host,
+        baseDomain
+      );
+    }
```

Comment:
great work! while you are here, any chance you want to take a crack at background-image, or other css assets?